File: CPP-Python-ROS-Apps/roscpp_demoapps/src/demoapp1.cpp
#include <ros/ros.h>
#include <core_api/ParamGetGlobalNamespace.h>
#include <core_api/TakeOff.h>
#include <core_api/Land.h>
#include <core_api/PositionSet.h>

std::string global_namespace;

core_api::ParamGetGlobalNamespace namespace_srv;
core_api::TakeOff takeoff_srv;
core_api::Land land_srv;
core_api::PositionSet pos_srv;

ros::ServiceClient pos_client,land_client,takeoff_client;


bool position_set(float x, float y, float z)
{
  pos_srv.request.twist.twist.linear.x = x;
  pos_srv.request.twist.twist.linear.y = y;
  pos_srv.request.twist.twist.linear.z = z;
  pos_srv.request.twist.twist.angular.z = 0.0;
  pos_srv.request.tolerance = 1.0;
  pos_srv.request.async = false;
  pos_srv.request.yaw_valid = false;
  pos_srv.request.relative = false;
  pos_srv.request.body_frame = false;
  ROS_INFO("Going to the next setpoint");
  pos_client.call(pos_srv);
  return pos_srv.response.success;
}

int main(int argc, char **argv)
{
  ros::init(argc, argv, "roscpp_demoapp1");
  ros::NodeHandle nh;

  ros::ServiceClient namespace_client = nh.serviceClient<core_api::ParamGetGlobalNamespace>("/get_global_namespace");
  namespace_client.call(namespace_srv);
  global_namespace = namespace_srv.response.param_info.param_value;
  
  takeoff_client = nh.serviceClient<core_api::TakeOff>("/"+global_namespace+"/navigation/take_off");
  land_client    = nh.serviceClient<core_api::Land>("/"+global_namespace+"/navigation/land");
  pos_client     = nh.serviceClient<core_api::PositionSet>("/"+global_namespace+"/navigation/position_set");

  ROS_INFO("Taking Off");
  takeoff_srv.request.takeoff_alt = 3.0;
  takeoff_client.call(takeoff_srv);
  if(!takeoff_srv.response.success)
  {
    ROS_ERROR("Failed to takeoff");
    return 1;    
  }
  // position_set(5,0,-3);       //Sending Position Setpoints
  // position_set(5,5,-3);
  // position_set(0,5,-3);
  // position_set(0,0,-3);

  if(!position_set(5,0,-3))
  {
    ROS_ERROR("Failed to set position");
    return 1;    
  }

  if(!position_set(5,5,-3))
  {
    ROS_ERROR("Failed to set position");
    return 1;    
  }

  if(!position_set(0,5,-3))
  {
    ROS_ERROR("Failed to set position");
    return 1;    
  }

  if(!position_set(0,0,-3))
  {
    ROS_ERROR("Failed to set position");
    return 1;    
  }

  ROS_INFO("Landing");
  land_srv.request.async =false;
  land_client.call(land_srv); 
  if(!land_srv.response.success)
  {
    ROS_ERROR("Failed to Land!");
    return 1;
  }
  return 0;
}

File: CPP-Python-ROS-Apps/ros_demoapps/src/demoapp1.py
#!/usr/bin/env python
import rospy
from core_api.srv import *

global_namespace = ''

def setpoint_local_position(lx, ly, lz, yaw=0.0, tolerance= 1.0, async = False, relative= False, yaw_valid= False, body_frame= False):
	global global_namespace
	rospy.wait_for_service('/'+ global_namespace +'/navigation/position_set')
	try:
		handle = rospy.ServiceProxy('/'+ global_namespace +'/navigation/position_set', PositionSet)

		# building message structure
		header_msg = std_msgs.msg.Header(1,rospy.Time(0.0,0.0),'a')
		twist = geometry_msgs.msg.Twist(geometry_msgs.msg.Vector3(lx,ly,lz),geometry_msgs.msg.Vector3(0.0,0.0,yaw))
		twiststamped_msg= geometry_msgs.msg.TwistStamped(header_msg, twist)
		req_msg = PositionSetRequest(x=lx, y=ly, z=lz, yaw=yaw, tolerance=tolerance, async=async, relative=relative, yaw_valid=yaw_valid, body_frame=body_frame)
		resp = handle(req_msg)
		return resp.success
	except rospy.ServiceException, e:
		rospy.logerr("pos set service call failed %s", e)
		return None

def make_square():
	global global_namespace
	#first get the global namespace to call the subsequent services
	#wait for service to come up
	rospy.wait_for_service('/get_global_namespace')
	try:
		res = rospy.ServiceProxy('/get_global_namespace', ParamGetGlobalNamespace)
		op = res()
		global_namespace = str(op.param_info.param_value)
	except rospy.ServiceException, e:
		rospy.logerr("global namespace service not available", e)
		#cannot continue without global namespace
		return None

	# Next take off to an altitue of 3.0 meters
	rospy.wait_for_service('/'+ global_namespace +'/navigation/take_off')
	try:
		handle = rospy.ServiceProxy('/'+ global_namespace +'/navigation/take_off', TakeOff)
		resp = handle(takeoff_alt=3.0)
	except rospy.ServiceException, e:
		rospy.logerr("takeoff service call failed %s", e)
		# cannot continue without taking off
		return None
	print "Took off successfully"

	# Then call the position set service for each edge of a square shaped trajectory
	if setpoint_local_position(5,0,-3.0):
		print "Successfully reached 1st waypoint"
	else:	
		rospy.logerr("Failed to set position")
		return None
	if setpoint_local_position(5,5,-3.0):
		print "Successfully reached 2nd waypoint"
	else:
		rospy.logerr("Failed to set position")
		return None
	if setpoint_local_position(0,5,-3.0):
		print "Successfully reached 3rd waypoint"
	else:
		rospy.logerr("Failed to set position")
		return None
	if setpoint_local_position(0,0,-3.0):
		print "Successfully reached 4th waypoint"
	else:
		rospy.logerr("Failed to set position")
		return None

	# Finally land the drone
	rospy.wait_for_service('/'+ global_namespace +'/navigation/land')
	try:
		handle = rospy.ServiceProxy('/'+ global_namespace +'/navigation/land', Land)
		resp = handle(False)
	except rospy.ServiceException, e:
		rospy.logerr("land service call failed %s", e)
		return None
	print "Landed Successfully. Exiting script."

if __name__ == "__main__":
	make_square()
